engine/common/stop/stop_control_test.go

```diff
@@ -0,0 +1,138 @@
+// 2. When a nil version is provided, it checks that the version data is cleared.
+func TestStopControl_OnVersionUpdate(t *testing.T) {
+	RunWithStopControl(t, func(_ *irrecoverable.MockSignalerContext, sc *StopControl) {
+
+		// Case 1: Version is updated
+		height := uint64(100)
+		version := semver.New("1.0.0")
+
+		sc.OnVersionUpdate(height, version)
+
+		// Verify that the version data is correctly stored.
+		versionData := sc.versionData.Load()
+		assert.NotNil(t, versionData)
+		assert.Equal(t, height, versionData.incompatibleBlockHeight)
+		assert.Equal(t, "1.0.0", versionData.updatedVersion)
+
+		// Case 2: Version update is deleted (nil version)
+		sc.OnVersionUpdate(0, nil)
+
+		// Verify that the version data is cleared.
+		versionData = sc.versionData.Load()
+		assert.Nil(t, versionData)
+	})
+}
+
+// TestStopControl_OnProcessedBlock tests the onProcessedBlock method of the StopControl.
+//
+// This test covers multiple scenarios related to processing block heights:
+// 1. Verifying that the processed height is updated correctly.
+// 2. Ensuring that a lower processed height cannot overwrite a higher one.
+// 3. Testing that the StopControl correctly triggers an irrecoverable error (via Throw) when the incompatible block height is reached.
+func TestStopControl_OnProcessedBlock(t *testing.T) {
+	RunWithStopControl(t, func(ctx *irrecoverable.MockSignalerContext, sc *StopControl) {
+		// Initial block height
+		height := uint64(10)
+
+		// Update processed height and verify it's stored correctly.
+		sc.updateProcessedHeight(height)
+		assert.Equal(t, height, sc.lastProcessedHeight.Value())
+
+		// Attempt to set a lower processed height, which should not be allowed.
+		sc.updateProcessedHeight(height - 1)
+		assert.Equal(t, height, sc.lastProcessedHeight.Value())
+
+		// Set version metadata with an incompatible height and verify the processed height behavior.
+		incompatibleHeight := uint64(13)
+		version := semver.New("1.0.0")
+
+		sc.OnVersionUpdate(incompatibleHeight, version)
+		height = incompatibleHeight - 2
+		sc.updateProcessedHeight(height)
+		assert.Equal(t, height, sc.lastProcessedHeight.Value())
+
+		// Prepare to trigger the Throw method when the incompatible block height is processed.
+		height = incompatibleHeight - 1
+
+		var wg sync.WaitGroup
+		wg.Add(1)
+
+		// Expected error message when the incompatible block height is processed.
+		expectedError := fmt.Errorf("processed block at height %d is incompatible with the current node version, please upgrade to version %s starting from block height %d", height, version.String(), incompatibleHeight-1)
+
+		// Set expectation that the Throw method will be called with the expected error.
+		ctx.On("Throw", expectedError).Run(func(args mock.Arguments) { wg.Done() }).Return().Once()
+
+		// Update the processed height to the incompatible height and wait for Throw to be called.
+		sc.updateProcessedHeight(height)
+		wg.Wait()
```

Comment:
use the `unittest.RequireReturnsBefore` method to timeout if wait is never reached